Add multiprofile password-autosave and general-autofill interfaces (#1854)
MicrosoftEdge/WebView2Feedback: specs/MultiProfile.md

with this profile. The cookie manager got from profile (CoreWebView2.Profile.CookieManager) is the
same as that got from CoreWebView2 (CoreWebView2.CookieManager).

Currently, we already have **ICoreWebView2Settings** interface to manage password-autosave and
general-autofill, but it will work not immediately but after the next navigation, and it can only
apply for current WebView2, which means if we start a new WebView2 using the same profile, all the
settings are default values and cannot be set from the profile. By adding password-autosave and
general-autofill management interfaces in profile, we can manage the properties and they will apply
immediately if we set a new value, and all WebView2s that created with the same profile can share
the settings, which means if we change password-autosave or general-autofill property in one WebView2,
the others with the same profile will also work. And these two property is linked with the same
properties in ICoreWebView2Settings, so changing one will change the other. So for the WebView2s
with the same profile, their **IsPasswordAutosaveEnabled** or **IsGeneralAutofillEnabled** property

> in **CoreWebView2Settings** and **CoreWebView2Profile** should always keep in sync.

Reviewer: This seems in conflict. You say the properties on the settings and the profile should always keep in-sync, but you also say the settings properties take effect on the next navigation and the profile settings take effect immediately. If the properties are applied at different times then it doesn't seem like they will always be in-sync. Or if they are always in-sync it doesn't seem like they can be applied at different times.

If you change the settings property, does the profile property only change after the settings property is applied during the next navigation? Or does the profile property change immediately when the settings property is changed? 

If they aren't always in-sync, which one 'wins' if both are set to different values?

Author: Good catch! Now the these two properties will work immediately no matter setting them in **CoreWebView2Settings** or **CoreWebView2Profile**, and when the property is changed in one interface, the same property in the other interface is changed as well immediately.

Reviewer: This is potentially a compat breaking behavior change: Modifying the existing properties to set immediately instead of next navigation. Although for these properties in particular I would guess its unlikely to practically be a big problem. Also, it looks like this is already checked in like this for a while so that's fine. We'll just need to also update the existing docs for those properties to say that they take effect immediately (unlike most other things in Settings)

Author: Make sense, will update the existing docs, thanks!

Reviewer: Yes, that's it. Please complete this PR, open a new PR to main, let me know, and I will schedule a review. Thanks!

    // ...
}
```

### Manage password-autosave and general-autofill settings in profile

```cpp
HRESULT AppWindow::ManagePasswordAutosaveInProfile(ICoreWebView2Controller* controller)
{
    // ...

    // Get the profile object.
    wil::com_ptr<ICoreWebView2> coreWebView2;
    CHECK_FAILURE(m_controller->get_CoreWebView2(&coreWebView2));
    Microsoft::WRL::ComPtr<ICoreWebView2Profile> webView2Profile;
    CHECK_FAILURE(webView2->get_Profile(&webView2Profile));
    
    // Get current value of password-autosave property.
    BOOL enabled;
    CHECK_FAILURE(webView2Profile->get_IsPasswordAutosaveEnabled(&enabled));

    // Set password-autosave property to the opposite value to current value.

>     if (enabled) {

Reviewer: ```suggestion
    if (enabled)
    {
```

        CHECK_FAILURE(webView2Profile->put_IsPasswordAutosaveEnabled(FALSE));
        MessageBox(
            nullptr, L"Password autosave will be disabled immediately in all WebView2 with the same profile.",
            L"Profile settings change", MB_OK);
    }

>     else {

Reviewer: ```suggestion
    else
    {
```

        CHECK_FAILURE(webView2Profile->put_IsPasswordAutosaveEnabled(TRUE));
        MessageBox(
            nullptr, L"Password autosave will be enabled immediately in all WebView2 with the same profile.",
            L"Profile settings change", MB_OK);
    }
  
    // ...
}

HRESULT AppWindow::ManageGeneralAutofillInProfile(ICoreWebView2Controller* controller)
{
    // ...

    // Get the profile object.
    wil::com_ptr<ICoreWebView2> coreWebView2;
    CHECK_FAILURE(controller->get_CoreWebView2(&coreWebView2));
    Microsoft::WRL::ComPtr<ICoreWebView2Profile> webView2Profile;
    CHECK_FAILURE(webView2->get_Profile(&webView2Profile));
    
    // Get current value of general-autofill property.
    BOOL enabled;
    CHECK_FAILURE(webView2Profile->get_IsGeneralAutofillsaveEnabled(&enabled));

    // Set general-autofill property to the opposite value to current value.

>     if (enabled) {

Reviewer: Curly braces go on their own line in Microsoft coding style guide
```suggestion
    if (enabled)
    {
```

        CHECK_FAILURE(webView2Profile->put_IsGeneralAutofillEnabled(FALSE));
        MessageBox(
            nullptr, L"General autofill will be disabled immediately in all WebView2 with the same profile.",
            L"Profile settings change", MB_OK);
    }

>     else {

Reviewer: ```suggestion
    else
    {
```

        String ProfilePath { get; };
    }

>     runtimeclass CoreWebView2Profile2 : CoreWebView2Profile

Reviewer: This isn't what it should look like for additions to a runtimeclass. It should look like a new [interface_name(...)] block inside the existing CoreWebView2Profile runtimeclass. If you run build-apiwriter.ps1 you can see your additions in the generated Microsoft.Web.WebView2.Core.idl file.

Author: Thanks Dave! It's the first time for me to know where the runtimeclass comes from. By running `powershell .\third_party\edge_webview2\win\webview2_api_writer\dotNetAPIWrapper\build-apiwriter.ps1 -BuildTarget WinRT -Configuration Debug -Architecture x64 -PackageKind Internal`, I got some code like below in Microsoft.Web.WebView2.Core.idl file:
```
runtimeclass CoreWebView2Profile
{
    // ICoreWebView2StagingProfile2 members
    Boolean IsPasswordAutosaveEnabled { get; set; };
    
    Boolean IsGeneralAutofillEnabled { get; set; };
    
    [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2ExperimentalProfile")]
    {
        // ICoreWebView2ExperimentalProfile members
        String ProfileName { get; };
    
        Boolean IsInPrivateModeEnabled { get; };
    
        String ProfilePath { get; };
    }
}
```
It seems that only `ProfileName`, `IsInPrivateModeEnabled` and `ProfilePath` are in the block `[interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2ExperimentalProfile")]`, while `IsPasswordAutosaveEnabled` and `IsGeneralAutofillEnabled` are outside any interface block. I'm not sure is there any problems about it, so may need your help to review and confirm. Thanks!

Reviewer: Oh oops, that's because of the interface isn't public yet. Once the experimental profile APIs become public and yours become public (which is what we want to show in these documents - what we expect it to look like when public) it should look like:
```
runtimeclass CoreWebView2Profile
{
    String ProfileName { get; };    
    Boolean IsInPrivateModeEnabled { get; };    
    String ProfilePath { get; };

    [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2Profile2")]
    {
        Boolean IsPasswordAutosaveEnabled { get; set; };
        Boolean IsGeneralAutofillEnabled { get; set; };    
    }
}
```

Author: Thanks for clarification! Document updated.

  /// See ICoreWebView2CookieManager.
  [propget] HRESULT CookieManager([out, retval] ICoreWebView2CookieManager** cookieManager);
}

[uuid(e2e8dce3-8213-4a32-b3b0-c80a8d154b61), object, pointer_default(unique)]

> interface ICoreWebView2Profile3 : ICoreWebView2Profile {

Reviewer: It should derive from latest interface version

```suggestion
interface ICoreWebView2Profile3 : ICoreWebView2Profile2 {
```

  /// IsPasswordAutosaveEnabled controls whether autosave for password
  /// information is enabled. The IsPasswordAutosaveEnabled property behaves
  /// independently of the IsGeneralAutofillEnabled property. When IsPasswordAutosaveEnabled is
  /// false, no new password data is saved and no Save/Update Password prompts are displayed.
  /// However, if there was password data already saved before disabling this setting,
  /// then that password information is auto-populated, suggestions are shown and clicking on
  /// one will populate the fields.
  /// When IsPasswordAutosaveEnabled is true, password information is auto-populated,
  /// suggestions are shown and clicking on one will populate the fields, new data
  /// is saved, and a Save/Update Password prompt is displayed.
  /// The default value is `FALSE`.

>   /// This property is linked with `CoreWebView2Settings.IsPasswordAutosaveEnabled`, so

Reviewer: ```suggestion
  /// This property has the same value as `CoreWebView2Settings.IsPasswordAutosaveEnabled`, and
```

>   /// changing one will change the other. And all WebView2s that created with the same

Reviewer: ```suggestion
  /// changing one will change the other. All `CoreWebView2`s with the same
```

>   /// `CoreWebView2Profile` will share this property, so for the WebView2s with the same

Reviewer: ```suggestion
  /// `CoreWebView2Profile` will share the same value for this property, so for the `CoreWebView2`s with the same
```

  /// profile, their `CoreWebView2Settings.IsPasswordAutosaveEnabled` and 

>   /// `CoreWebView2Profile.IsPasswordAutosaveEnabled` should always keep in sync.

Reviewer: ```suggestion
  /// `CoreWebView2Profile.IsPasswordAutosaveEnabled` will always have the same value.
```

  [propget] HRESULT IsPasswordAutosaveEnabled([out, retval] BOOL* value);
  /// Set the IsPasswordAutosaveEnabled property.
  [propput] HRESULT IsPasswordAutosaveEnabled([in] BOOL value);

  /// IsGeneralAutofillEnabled controls whether autofill for information
  /// like names, street and email addresses, phone numbers, and arbitrary input
  /// is enabled. This excludes password and credit card information. When
  /// IsGeneralAutofillEnabled is false, no suggestions appear, and no new information
  /// is saved. When IsGeneralAutofillEnabled is true, information is saved, suggestions
  /// appear and clicking on one will populate the form fields.
  /// The default value is `TRUE`.

>   /// This property is linked with `CoreWebView2Settings.IsGeneralAutofillEnabled`, so

Reviewer: ```suggestion
  /// This property has the same value as `CoreWebView2Settings.IsGeneralAutofillEnabled`, and
```

  /// changing one will change the other. And all WebView2s that created with the same

>   /// `CoreWebView2Profile` will share this property, so for the WebView2s with the same

Reviewer: ```suggestion
  /// `CoreWebView2Profile` will share the same value for this property, so for the `CoreWebView2`s with the same
```

  /// profile, their `CoreWebView2Settings.IsGeneralAutofillEnabled` and 

>   /// `CoreWebView2Profile.IsGeneralAutofillEnabled` should always keep in sync.

Reviewer: ```suggestion
  /// `CoreWebView2Profile.IsGeneralAutofillEnabled` will always have the same value.
```
